packages/flutter/test/cupertino/tab_scaffold_test.dart

```diff
@@ -463,6 +527,202 @@ void main() {
+          tabBuilder: (BuildContext context, int index) => const Placeholder()
+        ),
+      )
+    );
+    expect(find.text('Tab 1'), findsOneWidget);
+    expect(find.text('Tab 2'), findsOneWidget);
+    expect(find.text('Tab 3'), findsOneWidget);
+
+    controller.dispose();
+    controller = CupertinoTabController(initialIndex: 0);
+    await tester.pumpWidget(
+      CupertinoApp(
+        home: CupertinoTabScaffold(
+          tabBar: CupertinoTabBar(
+            items: List<BottomNavigationBarItem>.generate(2, tabGenerator),
+          ),
+          controller: controller,
+          tabBuilder: (BuildContext context, int index) => const Placeholder()
+        ),
+      )
+    );
+
+    // Should not crash here.
+    expect(find.text('Tab 1'), findsOneWidget);
+    expect(find.text('Tab 2'), findsOneWidget);
+    expect(find.text('Tab 3'), findsNothing);
+  });
+
+  testWidgets('Assert when current tab index >= number of tabs', (WidgetTester tester) async {
+    final CupertinoTabController controller = CupertinoTabController(initialIndex: 2);
+
+    try {
+      await tester.pumpWidget(
+        CupertinoApp(
+          home: CupertinoTabScaffold(
+            tabBar: CupertinoTabBar(
+              items: List<BottomNavigationBarItem>.generate(2, tabGenerator),
+            ),
+            controller: controller,
+            tabBuilder: (BuildContext context, int index) => Text('Different page ${index + 1}'),
+          ),
+        )
+      );
+    } on AssertionError catch (e) {
+      expect(e.toString(), contains('controller.index < tabBar.items.length'));
+    }
+
+    await tester.pumpWidget(
+      CupertinoApp(
+        home: CupertinoTabScaffold(
+          tabBar: CupertinoTabBar(
+            items: List<BottomNavigationBarItem>.generate(3, tabGenerator),
+          ),
+          controller: controller,
+          tabBuilder: (BuildContext context, int index) => Text('Different page ${index + 1}'),
+        ),
+      )
+    );
+
+    expect(tester.takeException(), null);
+
+    controller.index = 10;
+    await tester.pump();
+
+    final String message = tester.takeException().toString();
+    expect(message, contains('current index ${controller.index}'));
+    expect(message, contains('with 3 tabs'));
+  });
+
+  testWidgets('Current tab index cannot go below zero or be null', (WidgetTester tester) async {
+    void expectAssertionError(VoidCallback callback, String errorMessage) {
+      try {
+        callback();
+      } on AssertionError catch (e) {
+        expect(e.toString(), contains(errorMessage));
+      }
+    }
+
+    expectAssertionError(() => CupertinoTabController(initialIndex: -1), '>= 0');
+    expectAssertionError(() => CupertinoTabController(initialIndex: null), '!= null');
+
+    final CupertinoTabController controller = CupertinoTabController();
+
+    expectAssertionError(() => controller.index = -1, '>= 0');
+    expectAssertionError(() => controller.index = null, '!= null');
+  });
+
```

Comment:
Should include a test where two tab scaffolds share the same controller. Both scaffolds should reflect programmatic changes to the controller. Rebuilding with just one tab scaffold shouldn't affect the controller and subsequent controller changes shouldn't fail because the old listeners (the ones that belonged to the second scaffold) were correctly removed.